Add Eight-Bit-Addressing mode to I2CEEBlockDevice.

boomer41 opened the pull request:
### Summary of changes

When dealing with EEPROMs without a 16 bit adressing, the current
implementation does not work, as it writes a 16 bit address to the chip.
This may cause undefined behaviour.
This change adds a new constructor argument to enable this new eight-bit
mode. It defaults to false to not break existing code.
This constructor argument should actually never be necessary to manually
set, except when dealing with cheap devices.

An example for such a device is the Hua Hong NEC K24C02. [Datasheet.](https://datasheet.lcsc.com/szlcsc/1811091410_Hua-Hong-NEC-K24C02_C10518.pdf)

#### Impact of changes

Added a new optional argument at the end of all existing constructors.

#### Migration actions required

None, as the new constructor argument is optional and at the end.
The default value does not change functionality in existing code.

### Documentation

Online documentation for I2CEEBlockDevice should feature the new constructors.

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [X] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR

boomer41 committed:
Add Eight-Bit-Adressing mode to I2CEEBlockDevice.

When dealing with EEPROMs without a 16 bit adressing, the current
implementation does not work, as it writes a 16 bit address to the chip.
This may cause undefined behaviour.
This change adds a new constructor argument to enable this new eight-bit
mode. It defaults to false to not break existing code.
This constructor argument should actually never be necessary to manually
set, except when dealing with cheap devices.

VeijoPesonen reviewed (approved):
Looks good to me

SeppoTakalo reviewed (dismissed):
Is this implementation now only working for 2kb version of the chip?
But with the 8-bit address, there is 4kb and bigger versions of the chip, where the high bits of the address is muxed with the device address.

**Example: How A0,A1&A2 are used as a page addresses on bigger devices**
![image](https://user-images.githubusercontent.com/3104794/74652141-0dfb4b80-518e-11ea-94c5-bacdbd93668d.png)

Can you fix your implementation so that with the same API, we are able to support 2K, 4K, 8K and 16K versions of this chip?

boomer41 committed:
I2CEEBlockDevice: Add paging to eight bit mode

boomer41 committed:
Refactor paged_handler to directly give the paged device address.

boomer41 committed:
Correctly check return codes from bytewise write function of I2C.

boomer41 committed:
Preserve original size as it is modified inside the handler function.

The value is const reference bound inside do_paged and must not be
modified, it is used inside the paging loop.

boomer41 committed:
Use pageSize-constant instead of hardcoded value

boomer41 committed:
Fix astyle

0xc0170 commented on components/storage/blockdevice/COMPONENT_I2CEE/I2CEEBlockDevice.h:
```diff
@@ -169,7 +177,36 @@ class I2CEEBlockDevice : public BlockDevice {
+
     int _sync();
+
+    using paged_handler = std::function<int(const bd_addr_t &address, const bd_size_t &length, const uint8_t &deviceAddress)>;
```
can this be changed to use `Callback` object instead ? That is our provided std:function replacement.

SeppoTakalo reviewed (dismissed):
Thanks for the contribution, but I feel like this is now way too complex for what it tries to achieve.

By using the lambda functions, the FLASH usage increrased by almost 10 kB. I don't think this has to be so complex.

What is wrong with just:
```C++
if (_eight_bit_address_mode) {
    if (!_i2c->write(_i2c_addr | (char)((addr & 0x300) >> 8)) ||
            !_i2c->write((char)(addr & 0xff))) {
        return BD_ERROR_DEVICE_ERROR;
        }
} else {
    if (!_i2c->write(_i2c_addr | 0) ||
            !_i2c->write((char)(addr >> 8)) ||
            !_i2c->write((char)(addr & 0xff))) {
        return BD_ERROR_DEVICE_ERROR;
    }
}
```

Then, to simplify things, we could use just the single byte write and sequential read. This would allow maximum compatiblity with minimal code size.

boomer41 committed:
Remove overcomplicated code from I2CEEBlockDevice

boomer41 committed:
Fix typo and clarify documentation

kjbracey commented on components/storage/blockdevice/COMPONENT_I2CEE/I2CEEBlockDevice.h:
```diff
@@ -169,7 +177,19 @@ class I2CEEBlockDevice : public BlockDevice {
+     * @param address An address in the requested page.
+     * @return The device's I2C address for that page
+     */
+    uint8_t get_paged_device_address(const bd_addr_t &address);
```
There's no point in passing a `bd_addr_t` by reference - it will be inefficient if the call isn't inlined. Just pass by value.

kjbracey commented on components/storage/blockdevice/COMPONENT_I2CEE/I2CEEBlockDevice.h:
```diff
@@ -169,7 +177,19 @@ class I2CEEBlockDevice : public BlockDevice {
     uint32_t _size;
     uint32_t _block;
 
+    bool _address_is_eight_bit;
```
Can you put this next to `_i2c_addr` - that will pack the structure better.

kjbracey commented on components/storage/blockdevice/COMPONENT_I2CEE/I2CEEBlockDevice.h:
```diff
@@ -169,7 +177,19 @@ class I2CEEBlockDevice : public BlockDevice {
+
+    /**
+     * Gets the device's I2C address with respect to the requested page.
+     * When eight bit mode is disabled, this function is a noop.
```
`eight-bit` should be hyphenated as it's an adjectival phrase.

kjbracey commented on components/storage/blockdevice/COMPONENT_I2CEE/I2CEEBlockDevice.cpp:
```diff
@@ -164,3 +196,14 @@ const char *I2CEEBlockDevice::get_type() const
+        // The page will be bits 2-4 of the user defined addresses.
+        return this->_i2c_addr | ((address & 0x0700u) >> 7u);
+    }
+}
```
Missing newline

kjbracey commented on components/storage/blockdevice/COMPONENT_I2CEE/I2CEEBlockDevice.cpp:
```diff
@@ -58,48 +62,76 @@ int I2CEEBlockDevice::read(void *buffer, bd_addr_t addr, bd_size_t size)
     MBED_ASSERT(is_valid_read(addr, size));
 
-    _i2c->start();
+    auto *pBuffer = reinterpret_cast<char *>(buffer);
```
`static_cast` suffices here, but not that strongly against reinterpret.

kjbracey commented on components/storage/blockdevice/COMPONENT_I2CEE/I2CEEBlockDevice.cpp:
```diff
@@ -58,48 +62,76 @@ int I2CEEBlockDevice::read(void *buffer, bd_addr_t addr, bd_size_t size)
     // Check the addr and size fit onto the chip.
     MBED_ASSERT(is_valid_program(addr, size));
 
+    auto const *pBuffer = reinterpret_cast<char const *>(buffer);
```
Normal Mbed OS code style would put the `const`s on the front.

kjbracey commented on components/storage/blockdevice/COMPONENT_I2CEE/I2CEEBlockDevice.cpp:
```diff
@@ -164,3 +196,14 @@ const char *I2CEEBlockDevice::get_type() const
+uint8_t I2CEEBlockDevice::get_paged_device_address(const bd_addr_t &address)
+{
+    if (!this->_address_is_eight_bit) {
+        return this->_i2c_addr;
```
Explicit `this->` is odd, and doesn't match rest of the file.

boomer41 committed:
Fix Code style and reorder class members to get besser packing

SeppoTakalo commented on components/storage/blockdevice/COMPONENT_I2CEE/I2CEEBlockDevice.cpp:
```diff
@@ -58,48 +62,76 @@ int I2CEEBlockDevice::read(void *buffer, bd_addr_t addr, bd_size_t size)
-            !_i2c->write((char)(addr & 0xff))) {
-        return BD_ERROR_DEVICE_ERROR;
-    }
+    while (size > 0) {
```
Are you sure that the read operation have to be paged for some devices?

If tried with AT24MAC and read operation as simple as:
```C++
int I2CEEBlockDevice::read(void *buffer, bd_addr_t addr, bd_size_t size)
{
    // Check the address and size fit onto the chip.
    MBED_ASSERT(is_valid_read(addr, size));

    auto *pBuffer = reinterpret_cast<char *>(buffer);

    _i2c->start();

    if (1 != _i2c->write(get_paged_device_address(addr))) {
        return BD_ERROR_DEVICE_ERROR;
    }

    if (!_address_is_eight_bit && 1 != _i2c->write((char)(addr >> 8u))) {
        return BD_ERROR_DEVICE_ERROR;
    }

    if (1 != _i2c->write((char)(addr & 0xffu))) {
        return BD_ERROR_DEVICE_ERROR;
    }

    _i2c->stop();

    if (0 != _i2c->read(_i2c_addr, pBuffer, size)) {
        return BD_ERROR_DEVICE_ERROR;
    }

    return BD_ERROR_OK;
}
```

And it works.. For me, it seems that only the write operation have to be single byte or full page, but no more. Read operation can be as large as the whole device.

This seems to be true for at least AT24C and Microchip 24XX devices:
http://ww1.microchip.com/downloads/en/devicedoc/21203m.pdf

>> To provide sequential reads, the 24XX256 contains an internal address
pointer which is incremented by one at the completion
of each operation. This address pointer allows the
entire memory contents to be serially read during one
operation. The internal address pointer will
automatically roll over from address 7FFF to address
0000 if the master acknowledges the byte received
from the array address 7FFF

kjbracey replied:
In my experience, I've never seen paging limitation on reads.

boomer41 replied:
Unfortunately i can't check because I only have the 2k variants.
When you're sure there's no limitation, I'll remove that one.

SeppoTakalo reviewed (commented):
Code seems much simpler now.

I tried with following example:
```C++
int main()
{
    printf("i2cee test\n");
    I2CEEBlockDevice i2cee(D14, D15, 0xa0, 256, 1, 400000, true);
    i2cee.init();
    // Write "Hello World!" to the first block
    char buffer[256] = {0};
    strcat(buffer, __FILE__ " : " "Hello World. Build in " __DATE__ " " __TIME__);
    i2cee.program(buffer, 0, strlen(buffer)+1);
    memset(buffer, 0, 100);
    // Read back what was stored, read the whole device to buffer
    i2cee.read(buffer, 0, 256);
    printf("Content: %s\n", buffer);
    // Deinitialize the device
    i2cee.deinit();
}
```

The code seems to work OK when writing, but did not properly read. I'm not sure why. I suggest that you simplify the read operation to non-paged version, unless there is a clear requirement by some device.

kjbracey commented on components/storage/blockdevice/COMPONENT_I2CEE/I2CEEBlockDevice.cpp:
```diff
@@ -164,3 +196,15 @@ const char *I2CEEBlockDevice::get_type() const
+        return _i2c_addr | ((address & 0x0700u) >> 7u);
+    }
+}
+
```
Think you've gone overboard and now have an excess blank line :)

(Any Git diff should show both cases in red normally).

boomer41 replied:
Oops, I didn't know where you wanted the NL.
Where exactly did you want the NL?
Thanks.

kjbracey replied:
It's possible I'm being misled by GitHub's view, but I think originally you had an unterminated final `}` line. I think you went to the bottom and pressed "Enter" twice rather than once, so you now have an excess blank line.

kjbracey commented on components/storage/blockdevice/COMPONENT_I2CEE/I2CEEBlockDevice.cpp:
```diff
@@ -22,17 +22,21 @@ using namespace mbed;
     PinName sda, PinName scl, uint8_t addr,
-    bd_size_t size, bd_size_t block, int freq)
-    : _i2c_addr(addr), _size(size), _block(block)
+    bd_size_t size, bd_size_t block, int freq,
```
This doesn't directly relate to this PR, but this driver should now really have a constructor that takes `const i2c_pinmap_t &` and passes to `I2C` to offer the possibility to avoid pulling in pinmap tables.

It would be nice to add this, but I guess as a follow-up PR.

boomer41 replied:
Yes, I think this should be a separate PR as it actually has nothing to do with the 8bit-mode.

boomer41 committed:
Remove paging and sync from I2CEEBlockDevice::read(), as it is not needed

SeppoTakalo reviewed (approved):
Looks good.
Thank you for the contribution.
